File: Dynamic-Programming/53.Maximum-Subarray/maximumSubarray.js
var maxSubArray = function (nums) {
    let currentMax = nums[0];
    let max = nums[0];

    if (nums.length === 1) return max;

    for (let i = 1; i < nums.length; i++) {
        currentMax = Math.max(nums[i], currentMax + nums[i]);
        max = Math.max(max, currentMax);
    }

    return max;
};

/**
 * @param {number[]} nums
 * @return {number}
 */
 var maxSubArray = function(nums) {
    let curMax = nums[0];
    let output = nums[0];
    
    for (let i = 1; i < nums.length; i++) {
        curMax = Math.max(nums[i], curMax + nums[i]);
        output = Math.max(curMax, output);
    }
    
    return output;
};

File: Dynamic-Programming/53.Maximum-Subarray/maximumSubarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        cur_max = nums[0]
        output = nums[0]
        
        for i in range(1, len(nums)):
            cur_max = max(nums[i], nums[i] + cur_max)
            output = max(output, cur_max)
        
        
        return output